packages/addon-dev/sample-typescript-rollup.config.js

```diff
@@ -0,0 +1,113 @@
+  *  The command to build:
+  *
+  *    `rollup -c ./config/rollup.config.js`
+  */
+import path from 'path';
+
+import alias from '@rollup/plugin-alias';
+import multiInput from 'rollup-plugin-multi-input';
+import ts from 'rollup-plugin-ts';
+import { defineConfig } from 'rollup';
+import { nodeResolve } from '@rollup/plugin-node-resolve';
+import { Addon } from '@embroider/addon-dev/rollup';
+
+import babelConfig from './babel.config';
+import packageJson from '../package.json';
+
+const addon = new Addon({
+  srcDir: 'src',
+  destDir: 'dist',
+});
+
+const transpilation = [
+  // Instruct rollup how to resolve ts and hbs imports
+  // (importing a template-only component, for example)
+  nodeResolve({ resolveOnly: ['./'], extensions: ['.js', '.ts', '.hbs'] }),
+
+  // Allow top-level imports (what folks are used to from v1 addons)
+  // During the build, anything referencing a top-level import will be
+  // replaced with a relative import.
+  // DANGER: it's somewhat easy to cause circular references with this tool
+  alias({
+    entries: [
+      {
+        find: '#types',
+        replacement: path.resolve('src', '-private', 'types.ts'),
+      },
+      {
+        find: packageJson.name,
+        replacement: path.resolve('src'),
+      },
+      {
+        find: `${packageJson.name}/(.*)`,
+        replacement: path.resolve('src/$1'),
+      },
+    ],
+  }),
+
+  // This babel config should *not* apply presets or compile away ES modules.
+  // It exists only to provide development niceties for you, like automatic
+  // template colocation.
+  // See `babel.config.json` for the actual Babel configuration!
+  ts({
+    // can be changed to swc or other transpilers later
+    // but we need the ember plugins converted first
+    // (template compilation and co-location)
+    transpiler: 'babel',
+    browserslist: false,
+    babelConfig,
+    // setting this true greatly improves performance, but
+    // at the cost of safety (and no declarations output in your dist directory).
+    transpileOnly: false,
+    tsconfig: {
+      fileName: 'tsconfig.json',
+      hook: (config) => ({ ...config, declaration: true }),
+    },
+  }),
+
+  // Follow the V2 Addon rules about dependencies. Your code can import from
+  // `dependencies` and `peerDependencies` as well as standard Ember-provided
+  // package names.
+  addon.dependencies(),
+
+  // Ensure that standalone .hbs files are properly integrated as Javascript.
+  addon.hbs(),
+
+  // addons are allowed to contain imports of .css files, which we want rollup
+  // to leave alone and keep in the published output.
+  // addon.keepAssets(['**/*.css']),
+];
+
+// these should be JS, even though the authored format is TS
+const globallyAvailable = ['components/**/*.js', 'services/*.js'];
+
+export default defineConfig({
+  input: ['src/**/*{js,hbs,ts}'],
+  output: addon.output(),
+  plugins: [
+    // allows specifying glob inputs so we can transpile each module separately
+    multiInput(),
```

Comment:
Why did you need to include this plugin? Interestingly I had the same plugin in glimmer-apollo before moving to embroider addon-dev, which seem to solve the issue with `publicEntrypoints`.

Reply:
I couldn't _not_ need to specify inputs. Rollup wouldn't tell me why. But this enables each file to be transpiled separately so that the appReexports code here: https://github.com/embroider-build/embroider/blob/main/packages/addon-dev/src/rollup-app-reexports.ts#L21
Can find and re-export files correctly.

Previously, i'd combine all my components into a single file in dist, thinking it'd improve end-build time of the consuming app. But, fighting the tooling isn't worth it in this case.

If addon-dev had an api for specifying which file and which exports to use, that'd be cool. Like,
```js
// dist/_app_/components/my-component.js
export { MyComponent as default } from 'my-addon';
```

Reply:
@scalvert and I were just looking into this as well (trying to use addon-dev with a TS project), and are running into the "You must specify options.input" error you're referring to here. 

It is thrown here:

https://github.com/rollup/rollup/blob/e0b10605597ea2f7ee753e6f4e1ae152d19e6508/src/Graph.ts#L166-L179

Reply:
I also found an issue with multi-input, where it adds a bunch of extra imports to the output files.
(I updated the PR description with an example).

Over at ember-resources, https://github.com/NullVoxPopuli/ember-statechart-component/pull/244
I've gradually been tweaking the rollup config until the output looks more correct.
Downside, is that, with everything in one file, consumers will require tree-shaking in order to get rid of the utilities they don't use.

Reply:
I think all that might be required to avoid "You must specify options.input" is to make sure the `publicEntrypoints` plugin is renaming .ts inputs to .js outputs, so that there will be some .js chunks emitted to avoid that assertion.

Reply:
It seems like in a non-TypeScript project things work without setting `input` through this pathway:

https://github.com/rollup/rollup/blob/e0b10605597ea2f7ee753e6f4e1ae152d19e6508/src/rollup/rollup.ts#L114-L116
https://github.com/rollup/rollup/blob/e0b10605597ea2f7ee753e6f4e1ae152d19e6508/src/utils/options/normalizeInputOptions.ts#L49
https://github.com/rollup/rollup/blob/e0b10605597ea2f7ee753e6f4e1ae152d19e6508/src/utils/options/normalizeInputOptions.ts#L156-L159

Reply:
> Previously, i'd combine all my components into a single file in dist, thinking it'd improve end-build time of the consuming app. 

This is probably bad for final app optimization. While it's *possible* to tree shake inside modules, it's still a sketchy thing to do and if you ever want to be able to use native modules in the browser you shouldn't rely on it.

So shipping separate entrypoints per component is better, because then apps will only pull in the modules they actually use.

Rollup already does the right thing here by combining together modules that *definitely* get used together, while leaving apart modules that only *might* get used together.

Reply:
> So shipping separate entrypoints per component is better, because then apps will only pull in the modules they actually use.

for sure -- is the primary motivation for trying to figure this out atm.
it's just "something" is making compiling correctly "impossible" when TS gets involved.
(still a lot of unknowns from my perspective)
 - I no longer think multi-input is a good idea
 - but I don't have other options lol (yet!)

Reply:
glimmer-apollos already solved this problem of making publicEntrypoints work for typescript. We should upstream this: https://github.com/josemarluedke/glimmer-apollo/blob/6671a2191f2032cfe6e0cad018d9f66fbb10af75/packages/glimmer-apollo/rollup.config.js#L18